### PR TITLE
OAuth2 (Keycloak) Integration with a Toy Endpoint

### PR DESCRIPTION
**Pull request by Kanakanajm**
I didn't dare to touch any files so I defined classes methods in files that are already there. Most of my codes are under `api/src/api/security_api.py` which contains several request filter methods that extract and verify token from the request header and then cast it into a simple [`User`](https://github.com/SciCompMod/ESID-Backend/blob/cec29e6bc9a581b1439433ae4001a8842d32a6bf/api/src/api/services/auth.py#L4) object defined in `api/src/api/services/auth.py`. The toy endpoint is defined in `api/src/api/app/apis/simulations_api.py` as [`create_protected_scenario`](https://github.com/SciCompMod/ESID-Backend/blob/cec29e6bc9a581b1439433ae4001a8842d32a6bf/api/src/api/app/apis/simulations_api.py#L51).

To test it out, you need to switch to [my branch](https://github.com/DLR-SC/ESID/tree/feature/auth-backend-integration) for the ESID frontend. There I attached realm information (required in the backend for verifying the token) in the request header to the toy endpoint.

Cheers

### CODE REVIEW

**Review by JonasGilg — CHANGES_REQUESTED** [review]
Looks good and seems to be working. Can you update the `openapi3_0.yaml` to reflect the new toy endpoint?

File: api/src/api/app/db/models.py
[the remark this answers is not in the document]
Comment (Kanakanajm) [reply api/src/api/app/db/models.py:1]: I changed it and add a `owner` field before but then it seems I don't have to do that so I removed it.

[the remark this answers is not in the document]
Comment (Kanakanajm) [reply api/src/api/app/db/models.py:1]: > Looks good and seems to be working. Can you update the `openapi3_0.yaml` to reflect the new toy endpoint?

Shouldn't the schema be generated automatically by FastAPI after defining the endpoints? When I add my toy endpoint, it is visible under `/docs`. Besides, i don't think openapi-generator can generate any [dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/#what-is-dependency-injection) which the endpoint needs to filter out unauthorized users.

**Comment by Kanakanajm** [comment]
> Looks good and seems to be working. Can you update the `openapi3_0.yaml` to reflect the new toy endpoint?

Shouldn't the schema be generated automatically by FastAPI after defining the endpoints? When I add my toy endpoint, it is visible under `/docs`. Besides, i don't think openapi-generator can generate any [dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/#what-is-dependency-injection) which the endpoint needs to filter out unauthorized users.

**Comment by JonasGilg** [comment]
> > Looks good and seems to be working. Can you update the `openapi3_0.yaml` to reflect the new toy endpoint?
> 
> Shouldn't the schema be generated automatically by FastAPI after defining the endpoints? When I add my toy endpoint, it is visible under `/docs`. Besides, i don't think openapi-generator can generate any [dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/#what-is-dependency-injection) which the endpoint needs to filter out unauthorized users.

We actually generate the endpoints from the specification. Fast-API at the end shows the implemented endpoints on the /docs endpoint. In the end it should be identical to the specification.

**Review by JonasGilg — APPROVED** [review]

**Pull request merged**